docs/pages/access-controls/idps/saml-attribute-mapping.mdx

```diff
@@ -0,0 +1,205 @@
+    - `basic`: value equals to `urn:oasis:names:tc:SAML:2.0:attrname-format:basic`.
+
+Attribute mapping can be configured when adding a SAML application in Teleport web UI, or with `saml_idp_service_provider`
+resource spec created with `tctl create` or via API.
+```yaml
+kind: saml_idp_service_provider
+metadata:
+  name: example.com
+spec:
+  entity_id: https://example.com/saml/metadata
+  acs_url: https://example.com/saml/metadata
+  attribute_mapping:
+  - name: username
+    value: uid
+  - name: firstname
+    name_format: basic # optional, unspecified used as default if no value is provided.
+    value: user.spec.traits.firstname
+  - name: groups
+    name_format: urn:oasis:names:tc:SAML:2.0:attrname-format:basic # optional, full urn format.
+    value: user.spec.roles
+```
+
+## Predicate expressions
+
+Attribute values are authored using predicate expressions.
+
+When a service provider is created with attribute mapping, internally, the attribute mapping details (attribute name, name format and
+the corresponding value) are embedded as a SAML requested attribute elements in the service provider entity descriptor.
+
+Then, during an SSO request, SAML assertion service extracts the expressions from the entity descriptor and passes it to the
+predicate expression evaluator along with the authenticated user context.
+
+Once the expressions are evaluated, the resulting values are asserted in the SAML response under the requested attribute name.
+
+### Evaluation context
+
+The following user attributes are available for mapping between Teleport IdP and service providers:
+
+| Attributes | Syntax                                                       |
+| ---------- | ------------------------------------------------------------ |
+| Username   | `uid` or `user.metadata.name`.                               |
+| Roles      | `eduPersonAffiliation` or `user.spec.roles`.                 |
+| Traits     | `user.spec.traits.firstname`, `user.spec.traits.groups` etc. |
+
+Given a correct and supported predicate expression, attributes will be mapped as long as the requested attributes are present
+in Teleport.
+Attribute mapping that points to a non-existent value will not be included in SAML assertion.
+
+### Predicate expressions syntax
+
+Predicate expressions for attribute mapping are evaluated against user attributes that can be accessed using
+evaluation context listed above.
+
+The supported functions and methods are listed below, along with the usage syntax and it's result, evaluated 
+against the following reference user spec file:
+```yaml
+# reference user spec file
+kind: user
+metadata:
+  name: foobar
+spec:
+  roles:
+    - access
+    - editor
+    - dev-ssh
+  traits:
+    firstname:
+      - foo
+    lastname:
+      - BAR
+    displayname:
+      - foo bar
+    email:
+      - foobar@example.com
+    groups:
+      - okta-admin
+      - dev-sso
+      - dev-rdp
+```
+### Methods
```

Comment:
Given that we're using existing code for this bit, is there an existing doc page we could adapt for these methods?

Reply:
The underlying functions and methods are borrowed from `traits_map` from login rules. But 
1. Not all the methods are exposed/works in the context attribute mapping. Adding new `Dict` values is not possible due to the fact that the final attribute value can be either a string or array of string, but not `Pair` value. So we have to hand pick method reference.
2. The reference example shown in login rules page is specific to login rules. But here, I wanted to show examples that are somewhat copy-pastable for attribute mapping. Without attribute mapping specific example, users will have to go through lots of guess work to adapt login rules syntax. 

I have also made reference to the underlying semantics minimal so we can point to a canonical predicate expression page for detailed syntax in future. But the example have to be in this page.